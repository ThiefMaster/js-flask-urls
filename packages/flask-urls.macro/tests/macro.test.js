import macrosPlugin from 'babel-plugin-macros';
import pluginTester from 'babel-plugin-tester';
import urlMap from '../../../test-data/url-map';

pluginTester({
  plugin: macrosPlugin,
  pluginOptions: {
    flaskURLs: {
      urlMap,
    },
  },
  babelOptions: {filename: __filename},
  title: 'babel macro',
  tests: {
    // succesful cases
    'generates the function call': {
      code: `
        import flask from '../src/flask-urls.macro';
        const noParamsURL = flask\`no_params\`;
      `,
      snapshot: true,
    },
    'works with a custom import name': {
      code: `
        import kittens from '../src/flask-urls.macro';
        const noParamsURL = kittens\`no_params\`;
      `,
      snapshot: true,
    },
    'creates builder import only once': {
      code: `
        import flask from '../src/flask-urls.macro';
        const noParamsURL = flask\`no_params\`;
        const defaultParam = flask\`default_param\`;
      `,
      snapshot: true,
    },
    'rewrites import using a custom builder import location': {
      code: `
        import flask from '../src/flask-urls.macro';
        const noParamsURL = flask\`no_params\`;
      `,
      snapshot: true,
      pluginOptions: {
        flaskURLs: {
          urlMap,
          builder: 'myapp/flask-urls',
        },
      },
    },
    'rewrites import using a custom base path': {
      code: `
        import flask from '../src/flask-urls.macro';
        const noParamsURL = flask\`no_params\`;
      `,
      snapshot: true,
      pluginOptions: {
        flaskURLs: {
          urlMap,
          basePath: 'myapp/',
        },
      },
    },
    // error cases
    'fails with invalid endpoint': {
      code: `
        import flask from '../src/flask-urls.macro';
        flask\`test\`;
      `,
      error: 'flask-url.macro must reference a valid flask endpoint',
    },
    'fails with expressions': {
      code: `
        import flask from '../src/flask-urls.macro';
        flask\`\${foo}\`;
      `,
      error: 'flask-url.macro cannot contain expressions',
    },
    'fails if unused': {
      code: `
        import flask from '../src/flask-urls.macro';
      `,
      error: 'flask-urls.macro is imported but not used',
    },
    'fails with wrong macro usage': {
      code: `
        import flask from '../src/flask-urls.macro';
        flask();
      `,
      error: 'flask-urls.macro only supports tagged template expressions',
    },
    'fails with non-default import': {
      code: `
        import {other} from '../src/flask-urls.macro';
      `,
      error: 'flask-urls.macro requires a default import',
    },
    'fails with multiple import specifiers': {
      code: `
        import flask, {other} from '../src/flask-urls.macro';
      `,
      error: 'flask-urls.macro only supports a default import',
    },
  },
});
