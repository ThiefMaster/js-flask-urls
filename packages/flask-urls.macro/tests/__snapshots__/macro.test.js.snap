// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`babel macro creates builder import only once: creates builder import only once 1`] = `

import flask from '../src/flask-urls.macro';
const noParamsURL = flask\`no_params\`;
const defaultParam = flask\`default_param\`;

      ↓ ↓ ↓ ↓ ↓ ↓

import _buildFlaskURL from 'flask-urls';

const noParamsURL = _buildFlaskURL.bind(
  null,
  {
    endpoint: 'no_params',
    rules: [
      {
        args: [],
        converters: {},
        defaults: null,
        trace: [
          {
            data: '|',
            isDynamic: false,
          },
          {
            data: '/no',
            isDynamic: false,
          },
          {
            data: '/',
            isDynamic: false,
          },
        ],
      },
    ],
  },
  ''
);

const defaultParam = _buildFlaskURL.bind(
  null,
  {
    endpoint: 'default_param',
    rules: [
      {
        args: ['x'],
        converters: {},
        defaults: {
          x: 'nope',
        },
        trace: [
          {
            data: '|',
            isDynamic: false,
          },
          {
            data: '/def',
            isDynamic: false,
          },
          {
            data: '/',
            isDynamic: false,
          },
        ],
      },
      {
        args: ['x'],
        converters: {},
        defaults: null,
        trace: [
          {
            data: '|',
            isDynamic: false,
          },
          {
            data: '/def/v/',
            isDynamic: false,
          },
          {
            data: 'x',
            isDynamic: true,
          },
        ],
      },
    ],
  },
  ''
);


`;

exports[`babel macro generates the function call: generates the function call 1`] = `

import flask from '../src/flask-urls.macro';
const noParamsURL = flask\`no_params\`;

      ↓ ↓ ↓ ↓ ↓ ↓

import _buildFlaskURL from 'flask-urls';

const noParamsURL = _buildFlaskURL.bind(
  null,
  {
    endpoint: 'no_params',
    rules: [
      {
        args: [],
        converters: {},
        defaults: null,
        trace: [
          {
            data: '|',
            isDynamic: false,
          },
          {
            data: '/no',
            isDynamic: false,
          },
          {
            data: '/',
            isDynamic: false,
          },
        ],
      },
    ],
  },
  ''
);


`;

exports[`babel macro rewrites import using a custom base path: rewrites import using a custom base path 1`] = `

import flask from '../src/flask-urls.macro';
const noParamsURL = flask\`no_params\`;

      ↓ ↓ ↓ ↓ ↓ ↓

import _buildFlaskURL from 'flask-urls';

const noParamsURL = _buildFlaskURL.bind(
  null,
  {
    endpoint: 'no_params',
    rules: [
      {
        args: [],
        converters: {},
        defaults: null,
        trace: [
          {
            data: '|',
            isDynamic: false,
          },
          {
            data: '/no',
            isDynamic: false,
          },
          {
            data: '/',
            isDynamic: false,
          },
        ],
      },
    ],
  },
  'myapp/'
);


`;

exports[`babel macro rewrites import using a custom builder import location: rewrites import using a custom builder import location 1`] = `

import flask from '../src/flask-urls.macro';
const noParamsURL = flask\`no_params\`;

      ↓ ↓ ↓ ↓ ↓ ↓

import _buildFlaskURL from 'myapp/flask-urls';

const noParamsURL = _buildFlaskURL.bind(
  null,
  {
    endpoint: 'no_params',
    rules: [
      {
        args: [],
        converters: {},
        defaults: null,
        trace: [
          {
            data: '|',
            isDynamic: false,
          },
          {
            data: '/no',
            isDynamic: false,
          },
          {
            data: '/',
            isDynamic: false,
          },
        ],
      },
    ],
  },
  ''
);


`;

exports[`babel macro works in mock mode: works in mock mode 1`] = `

import flask from '../src/flask-urls.macro';
const fooURL = flask\`foo_not_defined\`;
const barURL = flask\`bar_not_defined\`;

      ↓ ↓ ↓ ↓ ↓ ↓

import {mockFlaskURL as _mockFlaskURL} from 'flask-urls';

const fooURL = _mockFlaskURL.bind(null, 'foo_not_defined');

const barURL = _mockFlaskURL.bind(null, 'bar_not_defined');


`;

exports[`babel macro works with a custom import name: works with a custom import name 1`] = `

import kittens from '../src/flask-urls.macro';
const noParamsURL = kittens\`no_params\`;

      ↓ ↓ ↓ ↓ ↓ ↓

import _buildFlaskURL from 'flask-urls';

const noParamsURL = _buildFlaskURL.bind(
  null,
  {
    endpoint: 'no_params',
    rules: [
      {
        args: [],
        converters: {},
        defaults: null,
        trace: [
          {
            data: '|',
            isDynamic: false,
          },
          {
            data: '/no',
            isDynamic: false,
          },
          {
            data: '/',
            isDynamic: false,
          },
        ],
      },
    ],
  },
  ''
);


`;
